services/offline/src/lib/dhis2-connection-status.tsx

```diff
@@ -0,0 +1,201 @@
+                    throw err
+                }
+            })
+            .then(() => {
+                // Ping is successful; set 'connected'
+                updateConnectedState(true)
+            })
+            .catch((err) => {
+                // Can get here if unauthorized, network error, etc.
+                console.error('Ping failed:', err.message)
+
+                // Only network errors should change status
+                const isNetworkErr = /^An unknown network error occurred$/.test(
+                    err.message
+                )
+                if (isNetworkErr) {
+                    updateConnectedState(false)
+                }
+            })
+    }, [engine, updateConnectedState])
+
+    /** Called when SW reports updates from incidental network traffic */
+    const onUpdate = useCallback(
+        ({ isConnected: newIsConnected }) => {
+            console.log('handling update from sw')
+            updateConnectedState(newIsConnected)
+            // Snooze ping timer to reduce pings since we know state from SW
+            smartIntervalRef.current?.snooze()
+        },
+        [updateConnectedState]
+    )
+
+    useEffect(() => {
+        const smartInterval = new SmartInterval({
+            initialDelay: 5000,
+            // don't ping if window isn't focused or visible
+            initialPauseValue:
+                !document.hasFocus() || document.visibilityState !== 'visible',
+            callback: ping,
+        })
+        smartIntervalRef.current = smartInterval
+
+        const handleBlur = () => smartInterval.pause()
+        const handleFocus = () => smartInterval.resume()
+        // On offline event, ping immediately to test server connection.
+        // Only do this when going offline -- it's theoretically no-cost
+        // for both online and offline servers. Pinging when going online
+        // can be costly for clients connecting over the internet to online
+        // servers.
+        const handleOffline = () => smartInterval.invokeCallbackImmediately()
+
+        window.addEventListener('blur', handleBlur)
+        window.addEventListener('focus', handleFocus)
+        window.addEventListener('offline', handleOffline)
+
+        return () => {
+            window.removeEventListener('blur', handleBlur)
+            window.removeEventListener('focus', handleFocus)
+            window.removeEventListener('offline', handleOffline)
+
+            // clean up smart interval
+            smartInterval.clear()
+        }
+    }, [ping])
+
+    useEffect(() => {
+        const unsubscribe = offlineInterface.subscribeToDhis2ConnectionStatus({
+            onUpdate,
+        })
+        return () => {
+            unsubscribe()
+        }
+    }, [offlineInterface, onUpdate])
+
+    // Memoize this value to prevent unnecessary rerenders of context provider
+    const contextValue = useMemo(
+        () => ({
+            isConnected,
+            isDisconnected: !isConnected,
```

Comment:
Related to my previous comment: since it turns our `isDisconnected` is simply the inverse of `isConnected` I suggest we remove it. The information is obsolete and may lead to confusion.